spec/middleware-1.0.0-draft.1.md

```diff
@@ -0,0 +1,309 @@
+the chain, and should return an Application.
+
+At run time, a Middleware can inspect and manipulate HTTP messages using the
+usual environment dictionary, and in by doing so leverages existing knowledge in
+building OWIN-compliant components.
+
+### 3.3. Signature
+
+```csharp
+using MidFunc = Func<
+                     AppFunc, // next Middleware in the chain
+                     AppFunc // returns the resulting application
+                    >
+```
+
+### 3.4. Code Sample
+
+The following example is a Middleware logging the request and response headers.
+
+```csharp
+// MidFunc
+public static AppFunc LogMiddleware(AppFunc nextMiddleware)
+{
+  AppFunc composedApplication = async environment => {
+
+    // log the request headers
+    Log.RequestHeaders(environment["owin.RequestHeaders"]);
+
+    // pass the request to the next Middleware
+    await next(environment);
+
+    // log the response headers
+    Log.RequestHeaders(environment["owin.ResponseHeaders"]);
+
+  }
+
+  return composedApplication;
+}
+```
+
+> **Note**
+>
+> This specification does not preclude alternative modeling of the Middleware
+> functionality from being provided
+> by vendors (e.g. object-oriented model, procedural, or separate before/after
+> functions), but they MUST provide a way to expose such alternate representations
+> as a MidFunc-compatible function.
+
+## 4. Middleware Registration
+
+The component responsible for producing the Application needs to be aware of all
+registered Middleware.
+
+In order to provide for capability discovery, configuration and shared
+initialisation, this specification introduces a factory function, taking as a
+parameter the Startup properties as defined by the
+[OWIN specification][owin-spec], and a builder function accepting that factory
+as a parameter.
+
+### 4.1. Signature
+
+```csharp
+using MidFactory = Func<
+                        IDictionary<string, object>, // startup properties
+                        MidFunc // returns the Middleware to use
+                       >;
+using BuildFunc = Action<
+                         MidFactory // the factory for a middleware
+                        >;
+```
+
+#### 4.1.1. Registration Sample
+
+The following example adds the `LogMiddleware` component defined earlier and
+registers it with the Builder.
+
+```csharp
+// as previously defined
+public static AppFunc LogMiddleware(AppFunc next) { /*  ... */ }
+
+// Definition of the factory
+public static MidFunc ConfigureLogMiddleware(IDictionary<string, object> startupProperties)
+{
+  // optionally read and write to the startup properties
+  return LogMiddleware;
+}
+```
+
+### 4.2. Discovery and Registration of Middleware
+
+A common requirement for Middleware authors is to enable users to discover
+Middleware after installing a software package in their Application.
+
+In order to deliver this functionality, a Middleware MUST implement an extension
```

Comment:
MUST? That seems like a strong requirement for an aesthetic component. I'd think this would be a SHOULD.